File: src/workflow/workflowSpec.test.ts
import { assertEquals, assertStringIncludes } from "jsr:@std/assert";
import {
  buildWorkflowGenerationSpec,
  documentedWorkflowNodeTypes,
  getWorkflowNodeSpec,
  getWorkflowSpecTool,
} from "./workflowSpec.ts";
import { workflowNodeTypes } from "./types.ts";

Deno.test("workflow spec tool returns requested node documentation and configured context", () => {
  assertEquals(getWorkflowSpecTool.name, "get_workflow_spec");
  const result = getWorkflowNodeSpec('["command", "http"]', {
    models: ["gemini-3.5-flash"],
    ragSettings: ["docs"],
    mcpServers: ["browser"],
  });
  assertStringIncludes(result, "- command:");
  assertStringIncludes(result, "- http:");
  assertStringIncludes(result, "gemini-3.5-flash");
  assertEquals(result.includes("- note:"), false);
});

Deno.test("workflow spec documents every executable node type", () => {
  assertEquals(documentedWorkflowNodeTypes(), [...workflowNodeTypes]);
  for (const type of workflowNodeTypes) {
    assertEquals(
      getWorkflowNodeSpec([type]).includes("unknown node type"),
      false,
      type,
    );
  }
});

Deno.test("workflow generation distinguishes caller input from interactive input", () => {
  const spec = getWorkflowNodeSpec();
  assertStringIncludes(spec, "It does not show an input dialog");
  assertStringIncludes(
    spec,
    "acquire every required user value with prompt-value",
  );
});

Deno.test("workflow generation requires an image model for image output", () => {
  assertStringIncludes(
    getWorkflowNodeSpec(["command"]),
    "saveImageTo requires an explicitly selected image-generation model",
  );
});

Deno.test("workflow generation treats infographics as structured text by default", () => {
  const spec = getWorkflowNodeSpec();
  assertStringIncludes(
    spec,
    'Interpret "infographic" as a readable, visually structured Markdown or HTML document by default',
  );
  assertStringIncludes(
    spec,
    "Do not assume it means a bitmap image",
  );
});

Deno.test("workflow generation spec includes runtime names and correctness examples", () => {
  const spec = buildWorkflowGenerationSpec({
    models: ["configured-model"],
    ragSettings: ["docs"],
    mcpServers: ["Browser"],
  });
  assertStringIncludes(spec, "source: responseBody");
  assertStringIncludes(spec, 'const text = "{{value:json}}"');
  assertStringIncludes(spec, "throwOnError: false");
  assertStringIncludes(spec, "Build the smallest connected workflow");
  assertStringIncludes(spec, "ragSetting: __websearch__");
  assertStringIncludes(spec, "http: reserved for APIs, webhooks");
  assertStringIncludes(spec, "Keep the YAML as concise and simple as possible");
  assertStringIncludes(spec, "configured-model");
  assertStringIncludes(spec, "Configured RAG settings: docs");
  assertStringIncludes(spec, "Enabled MCP servers");
});

File: workflow_spec_test.go
package main

import (
	"strings"
	"testing"
)

func TestWorkflowSpecToolFiltersNodeTypes(t *testing.T) {
	result := workflowSpecToolResult(map[string]any{"nodeTypes": []any{"command", "http"}})
	text, _ := result["result"].(string)
	if !strings.Contains(text, "- command:") || !strings.Contains(text, "- http:") || strings.Contains(text, "- note:") {
		t.Fatalf("unexpected filtered workflow spec: %s", text)
	}
	if !strings.Contains(text, "*.workflow.yaml") || strings.Contains(text, "fenced YAML") {
		t.Fatalf("workflow spec did not describe the canonical YAML file format: %s", text)
	}
}

func TestWorkflowSpecToolIncludesConfiguredContext(t *testing.T) {
	result := workflowSpecToolResult(map[string]any{"nodeTypes": []any{"command"}}, WorkflowSpecContext{Models: []string{"gemini-3.5-flash"}, RAGSettings: []string{"docs"}, MCPServers: []string{"browser"}})
	text, _ := result["result"].(string)
	for _, expected := range []string{"Configured models: gemini-3.5-flash", "Configured RAG settings: docs", "Configured MCP servers: browser"} {
		if !strings.Contains(text, expected) {
			t.Fatalf("missing %q in %s", expected, text)
		}
	}
}
